add buffersize argument to rdfind

trollkarlen opened the pull request:
add the argument to change buffer size of checksum calculation, may increase speed on some checksum algorithms and disk types.

pauldreik commented on Fileinfo.cc:
```diff
@@ -80,7 +81,7 @@ Fileinfo::fillwithbytes(enum readtobuffermode filltype,
     Checksum chk(checksumtype);
 
-    char buffer[4096];
+    char buffer[buffersize];
```
this is a variable length array - not standard C++.

I think it makes sense to only allocate it once. This is C++17, so we can't use std::span.

Please make a std::vector outside the loop, sized to buffersize and pass it by reference into this function.

trollkarlen replied:
I really like having this on the stack because of ease and also if this becomes multi-threaded sometime somewhere in the future. 
I atleast changed to std::vector, but c++ is not my stong side so please review carefully :)

pauldreik commented on rdfind.cc:
```diff
@@ -109,6 +113,7 @@ struct Options
   bool usesha1 = false;      // use sha1 checksum to check for similarity
   bool usesha256 = false;    // use sha256 checksum to check for similarity
   bool usesha512 = false;    // use sha512 checksum to check for similarity
+  long buffersize = 4096;    // chunksize to use when reading files
```
std::size_t

pauldreik commented on rdfind.cc:
```diff
@@ -184,6 +189,14 @@ parseOptions(Parser& parser)
       }
+    } else if (parser.try_parse_string("-buffersize")) {
+      const long buffersize = std::stoll(parser.get_parsed_string());
+      if (buffersize < 0) {
+        throw std::runtime_error("negative value of buffersize not allowed");
+      } else if (buffersize > 1000*1024) {
+        throw std::runtime_error("maximum 1000KB value of buffersize not allowed");
+      }
+      o.buffersize = buffersize;
```
please write to std::cerr and exit with failure like on rows 188-190

pauldreik reviewed (changes requested):
Stretch goal:

I am thinking of what the optimal buffersize is. It has been 19 years since I wrote rdfind, and hardware has changed quite a bit since then.
It would be good to have data on
 - a fast typical desktop (amd64 cpu, fast ssd)
 - a smaller system (raspberry pi3, sd card and 32 bit?)
 - a server (pretty similar to the desktop I assume, but storage is remote)

also, what is good for small files and what is good for large files?

pauldreik commented:
also, the title of the MR does not say it is rdfind, that is pretty implicit :smiley:

pauldreik commented on rdfind.cc:
```diff
@@ -65,6 +65,10 @@ usage()
        "device and inode\n"
     << " -checksum           md5 |(sha1)| sha256 | sha512\n"
     << "                                  checksum type\n"
+    << " -buffersize N     (N=4096,0<N<1000KB)\n"
```
kB

trollkarlen replied:
Form my understanding kB means 1000Bytes and KB means 1024Bytes, and in this case Im referring to 1024Bytes.
But i guess just saying this can restart the vim/emacs war again :)

pauldreik commented:
forgot: please add a test

trollkarlen commented:
Most should be fixed

trollkarlen commented:
had some error in sh script.
Added shellsheck workflow and script, just added the new files. 
Too many issues in the other files to fix here.
Maybe will do PR later to fix this.

trollkarlen commented:
I removed "add shellcheck script and workflow", from this patchset and will move that to new PR that fixes all issues aswell

trollkarlen commented:
Please review again it that changes is sufficient.

pauldreik commented:
I am sorry - I had made review comments but forgot to actually send them! On top of that, I missed I had to approve the CI runs.

I worked a bit on this to get this over the finish line. I made a branch and pushed it here. 

Most of the changes are things you could not reasonably know. Such things take long time to discuss over the internet, had it been at work we would just had a discussion and get it done quickly.

The differences:

1. set the default to 1 MiB (after measuring on a small device, after this it got slower. on a big device, there were improvements but very small)
2. fix reallocating the vector for each file
3. fix spelling and wording in the documentation
4. up the limit to 128 MiB, it is heap allocated so no problem with big values.
5. modified the speed test a bit to include large and small files and make the output more readable
6. moved the size member in the struct to avoid padding
7. removed the defaults from fillwithbytes, that was historical garbage
8. got rid of an include file (old sins)

I hope you are ok with the changes, I made them as fixups to your commits so you are still the author. Please let me know if you want something else!

see the branch here: https://github.com/pauldreik/rdfind/pull/180

pauldreik commented:
closing, in favour of #180
